File: tools/lib/zeo_dev/commands/diff.rb
# frozen_string_literal: true

require "fileutils"
require "tmpdir"

module ZeoDev
  module Commands
    # Run a ruby snippet through BOTH engines, show each side, and say whether
    # they agree. On a DIVERGENCE it captures the snippet as an XFAIL gap under
    # `tests/gaps/`, blessed from ruby; on a MATCH it writes nothing.
    #
    # This is the long-tail loop in one command: find a divergence, file it
    # with its golden, confirm the harness agrees it is real.
    #
    # The gaps harness is the ARBITER of whether a divergence is a real gap.
    # It normalizes source paths, so two outputs that differ here can still
    # match there -- after writing and blessing, the gap is run, and a
    # "GAP FIXED" verdict means the file is removed and the snippet belongs in
    # `tests/` instead.
    class Diff < Cli
      def self.summary = "compare a snippet across ruby and zeo, and file a gap"

      def self.banner = <<~TEXT
        usage: zeo-dev diff [<code> | -f <file> | -]

          zeo-dev diff 'puts 1 + 1'          # inline code
          zeo-dev diff -f snippet.rb         # from a file
          pbpaste | zeo-dev diff -           # from stdin
      TEXT

      def defaults = { name: "gap_snippet", gap: true, file: nil, stdin: false }

      def options(o)
        o.on("--name NAME", "-n NAME", "gap stem to use on divergence") { |v| opts[:name] = v }
        o.on("--no-gap", "compare only; never write a gap file") { opts[:gap] = false }
        o.on("-f FILE", "read the snippet from FILE") { |v| opts[:file] = v }
      end

      def run
        source = read_source
        raise Error, self.class.banner if source.nil?

        puts "=== snippet ==="
        puts source
        puts

        ruby = Exec.run(Ruby.oracle_argv("-"), env: Ruby.oracle_env, stdin: source, capture_stdout: true, chdir: ROOT)
        zeo = Exec.run([Ruby.build_zeo!, "-W0", "-e", source], capture_stdout: true, chdir: ROOT)
        show("ruby", ruby)
        puts
        show("zeo", zeo)
        puts

        # A quick verdict on stdout plus exit; the harness reconciles the fine
        # print.
        if ruby.stdout == zeo.stdout && ruby.code == zeo.code
          puts "MATCH (stdout + exit). Not a gap -- belongs in tests/ if you want to keep it."
          return 0
        end
        puts "DIVERGE:"
        puts "  ruby exit #{ruby.code.inspect}, zeo exit #{zeo.code.inspect}" if ruby.code != zeo.code
        show_stdout_diff(ruby.stdout, zeo.stdout)
        puts
        return puts("(--no-gap: not writing a gap file)") || 0 unless opts[:gap]

        write_gap(source)
      end

      private

      def read_source
        return File.read(opts[:file]) if opts[:file]
        return $stdin.read if args.first == "-" || (args.empty? && !$stdin.tty?)

        args.shift&.then { |s| "#{s}\n" }
      end

      def show(label, res)
        puts "--- #{label} (exit #{res.code.inspect}) ---"
        puts "stdout:"
        print res.stdout
        return if res.stderr.to_s.empty?

        puts "stderr:"
        print res.stderr
      end

      def show_stdout_diff(a, b)
        al = a.to_s.lines
        bl = b.to_s.lines
        [al.size, bl.size].max.times do |i|
          next if al[i] == bl[i]

          puts "  ruby #{i + 1}: #{al[i].inspect}"
          puts "  zeo  #{i + 1}: #{bl[i].inspect}"
        end
      end

      def write_gap(source)
        name = opts[:name].delete_suffix(".rb")
        dest = File.join(ROOT, "tests", "gaps", "#{name}.rb")
        raise Error, "refusing to overwrite tests/gaps/#{name}.rb (use --name)" if File.exist?(dest)

        File.write(dest, source)
        bless_gap(name)
        puts "wrote tests/gaps/#{name}.rb (+ blessed golden from ruby)"

        puts "confirming the XFAIL holds via the gaps harness ..."
        run_out = Exec.run(%W[cargo nextest run -p zeo --test goldens -E test(#{name})],
                           chdir: ROOT, capture_stdout: true)
        if "#{run_out.stdout}#{run_out.stderr}".include?("GAP FIXED")
          puts
          puts "the gaps harness says zeo actually MATCHES ruby (under source-path"
          puts "normalization) -- this is NOT a gap. Removing it; promote to tests/ instead."
          %w[rb rb.expected rb.err.expected].each do |suf|
            FileUtils.rm_f(File.join(ROOT, "tests", "gaps", "#{name}.#{suf}"))
          end
          return 1
        end
        puts "gap tests/gaps/#{name}.rb is a valid XFAIL (zeo diverges). Grind it down, then:"
        puts "  tools/zeo-dev promote-gap #{name}"
        0
      end

      # Delegates to `zeo-dev bless` -- the ONE golden writer -- so the
      # recorded output goes through the harness's own normalization (CRLF,
      # source-path relativization, address scrubbing). A hand-rolled
      # oracle capture here once skipped the address scrub, so a snippet
      # printing `#<Object:0x...>` recorded a raw process-random address
      # into its golden.
      def bless_gap(name)
        res = Exec.run([File.join(ROOT, "tools", "zeo-dev"), "bless", name],
                       chdir: ROOT, capture_stdout: true)
        raise Error, "bless failed for #{name}:\n#{res.stderr}" unless res.success?
      end
    end
  end
end

File: crates/xtask/src/commands/diff.rs
//! Run a ruby snippet through BOTH engines, show each side, and say whether
//! they agree. On a DIVERGENCE it files the snippet under `todo/` with
//! ruby's answer recorded under `__END__`; on a MATCH it writes nothing.
//!
//! This is the long-tail loop in one command: find a divergence, file it
//! with the answer zeo has to reach.
//!
//! `todo/` is not run by anything. When the program starts matching, it
//! moves into the topic directory it belongs to under `test/` and becomes an
//! ordinary test.

use std::io::Read;

use crate::exec::{self, Capture};
use crate::ruby::Oracle;
use crate::{Error, root, root_join};

const USAGE: &str = "\
usage: cargo xtask diff [<code> | -f <file> | -]

  cargo xtask diff 'puts 1 + 1'          # inline code
  cargo xtask diff -f snippet.rb         # from a file
  pbpaste | cargo xtask diff -           # from stdin
  --name NAME / -n NAME                  # gap stem to use on divergence
  --no-gap                               # compare only; never write a gap
";

pub fn run(args: &[String]) -> Result<(), Error> {
    let mut name = "gap_snippet".to_string();
    let mut gap = true;
    let mut file = None;
    let mut inline = None;
    let mut from_stdin = false;
    let mut rest = args.iter();
    while let Some(arg) = rest.next() {
        let mut value = |flag: &str| {
            rest.next()
                .cloned()
                .ok_or_else(|| Error::new(format!("{flag} wants a value\n\n{USAGE}")))
        };
        match arg.as_str() {
            "--name" | "-n" => name = value("--name")?,
            "-f" => file = Some(value("-f")?),
            "--no-gap" => gap = false,
            "-" => from_stdin = true,
            "--help" | "-h" => {
                print!("{USAGE}");
                return Ok(());
            }
            other if other.starts_with("--") => {
                return Err(Error::new(format!("unknown option {other:?}\n\n{USAGE}")));
            }
            other => inline = Some(other.to_string()),
        }
    }

    let source = read_source(file.as_deref(), inline.as_deref(), from_stdin)?
        .ok_or_else(|| Error::new(USAGE.to_string()))?;
    println!("=== snippet ===");
    print!("{source}");
    println!();

    let oracle = Oracle::find();
    let ruby = exec::run_with_stdin(
        &oracle.argv(&["-"]),
        root(),
        &oracle.env(),
        Capture::Both,
        Some(source.as_bytes()),
    )?;
    let zeo_bin = crate::build_zeo()?;
    let zeo = exec::run(
        &[
            zeo_bin.as_os_str(),
            std::ffi::OsStr::new("-W0"),
            std::ffi::OsStr::new("-e"),
            std::ffi::OsStr::new(&source),
        ],
        root(),
        &[],
        Capture::Both,
    )?;
    show("ruby", &ruby);
    println!();
    show("zeo", &zeo);
    println!();

    // A quick verdict on stdout plus exit; the harness reconciles the fine
    // print.
    if ruby.stdout == zeo.stdout && ruby.code == zeo.code {
        println!(
            "MATCH (stdout + exit). Not a gap -- belongs in a test/ topic dir if you want to keep it."
        );
        return Ok(());
    }
    println!("DIVERGE:");
    if ruby.code != zeo.code {
        println!(
            "  ruby exit {}, zeo exit {}",
            ruby.code_text(),
            zeo.code_text()
        );
    }
    show_stdout_diff(&ruby.stdout_text(), &zeo.stdout_text());
    println!();
    if !gap {
        println!("(--no-gap: not writing a gap file)");
        return Ok(());
    }
    write_gap(&name, &source)
}

fn read_source(
    file: Option<&str>,
    inline: Option<&str>,
    from_stdin: bool,
) -> Result<Option<String>, Error> {
    if let Some(path) = file {
        return std::fs::read_to_string(path)
            .map(Some)
            .map_err(|e| Error::new(format!("reading {path}: {e}")));
    }
    if from_stdin {
        let mut buf = String::new();
        std::io::stdin()
            .read_to_string(&mut buf)
            .map_err(|e| Error::new(format!("reading stdin: {e}")))?;
        return Ok(Some(buf));
    }
    Ok(inline.map(|s| format!("{s}\n")))
}

fn show(label: &str, out: &exec::Output) {
    println!("--- {label} (exit {}) ---", out.code_text());
    println!("stdout:");
    print!("{}", out.stdout_text());
    if out.stderr.is_empty() {
        return;
    }
    println!("stderr:");
    print!("{}", out.stderr_text());
}

fn show_stdout_diff(a: &str, b: &str) {
    let al: Vec<&str> = a.lines().collect();
    let bl: Vec<&str> = b.lines().collect();
    for i in 0..al.len().max(bl.len()) {
        if al.get(i) == bl.get(i) {
            continue;
        }
        println!("  ruby {}: {:?}", i + 1, al.get(i));
        println!("  zeo  {}: {:?}", i + 1, bl.get(i));
    }
}

fn write_gap(name: &str, source: &str) -> Result<(), Error> {
    let name = name.trim_end_matches(".rb");
    let dest = root_join("todo").join(format!("{name}.rb"));
    if dest.exists() {
        return Err(Error::new(format!(
            "refusing to overwrite todo/{name}.rb (use --name)"
        )));
    }
    std::fs::write(&dest, source)
        .map_err(|e| Error::new(format!("writing {}: {e}", dest.display())))?;
    // Through `bless` -- the ONE writer of a recorded answer -- so the output
    // gets the same normalization a test's does (CRLF, source-path
    // relativization, address scrubbing). A hand-rolled oracle capture here
    // once skipped the address scrub, so a snippet printing `#<Object:0x...>`
    // recorded a raw process-random address.
    super::bless::bless(&[&format!("todo::{name}.rb")], false)?;
    println!("wrote todo/{name}.rb, with ruby's answer under __END__");
    println!();
    println!("Nothing runs it. When zeo matches, move it into a topic dir:");
    println!("  git mv todo/{name}.rb test/<topic>/<area>/{name}.rb");
    Ok(())
}
